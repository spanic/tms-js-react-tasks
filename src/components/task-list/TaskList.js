import { useEffect } from 'react';

import './TaskList.scss';

function TaskList() {
  useEffect(() => {
    /**
     * Getting tasks from the server
     */
    getTasks();

    /**
     * "Delete all" button
     */
    const deleteAllButton = document.createElement('button');
    deleteAllButton.className = 'task-list__button';
    const deleteAllButtonText = document.createTextNode('Delete all');

    deleteAllButton.append(deleteAllButtonText);
    const firstControlsRow =
      document.getElementsByClassName('task-list__row')[0];
    firstControlsRow.prepend(deleteAllButton);

    /**
     * "Enter your task here..." input
     */
    // ...

    /**
     * "Total:" div
     */
    // ...

    /**
     * "Show completed" button
     */
    // ...

    /**
     * Add new task
     */
    // ...

    /**
     * Delete all tasks
     */
    // ...

    /**
     * Delete task
     */
    // ...

    /**
     * Complete task
     */
    // ...
  }, []);

  return (
    <div id="task-list" className="task-list">
      <div className="task-list__row">
        {/* Add it manually: */}
        {/* <button className="task-list__button">Delete all</button> */}
        <button className="task-list__button">Delete last</button>
        {/* Add it manually: */}
        <input
          id="taskTextInput"
          className="task-list__input"
          type="text"
          placeholder="Enter your task here..."
        ></input>
        <button id="addBtn" className="task-list__button">
          Add
        </button>
      </div>
      <div className="task-list__row">
        {/* Add it manually: */}
        <div>
          Total: <span>2</span>
        </div>
        <div>
          Completed: <span>1</span>
        </div>
        <button className="task-list__button">Show all</button>
        {/* Add it manually: */}
        <button className="task-list__button" disabled>
          Show completed
        </button>
        <input
          className="task-list__input"
          type="search"
          placeholder="Search..."
          disabled
        ></input>
      </div>
      <div className="task-list__container"></div>
    </div>
  );
}

function getTasks() {
  // ...
}

function addTask(text, animated = false) {
  const taskListContainer = document.querySelector('div.task-list__container');
  taskListContainer.insertAdjacentHTML(
    'afterbegin',
    `
    <div class="task-list__item task-list-item ${
      animated ? 'task-list__item_animated' : ''
    }">
      <div class="task-list-item__chevron"></div>
      <span class="task-list-item__text">
        ${text}
      </span>
      <div class="task-list-item__controls">
        <div class="task-list__row">
          <button class="task-list__button task-list__button_complete">
            Complete
          </button>
          <button class="task-list__button task-list__button_remove"></button>
        </div>
        <span class="task-list-item__created-date">
          02.03.2023 00:13 GMT+6
        </span>
      </div>
    </div>
    `
  );
}

export default TaskList;
