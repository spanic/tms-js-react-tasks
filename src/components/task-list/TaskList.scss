@use '../shared/styles/shared';

$royal-blue-traditional: #0a2463ff;
$imperial-red: #fb3640ff;
$davys-gray: #605f5eff;
$cerulean: #247ba0ff;
$platinum: #e2e2e2ff;
$magestic-green: #2f712f;
$light-dimmed-green: #b5d2b5;
$control-color-black: #000000;
$control-color-gray: #e8e8e8;

$control-border-radius: 0.75em;
$control-box-shadow-default-size: 0 0.23em 0 0;
$control-box-shadow-hover-size: 0 0.33em 0 0;
$control-default-translation: translateY(-0.2em);
$control-hover-translation: translateY(-0.33em);
$control-active-translation: translateY(0);

$task-item-fade-in-fade-out-animation-length: 0.35s;

@mixin control-default-shared-style {
  transition: transform 0.1s ease, box-shadow 0.1s ease;
  box-sizing: border-box;
  border: 2px solid $control-color-black;
  border-radius: $control-border-radius;
  padding: 0.5em 0.75em;

  &:focus,
  &:focus-visible {
    outline: none;
  }

  &:disabled {
    &,
    &:hover:not(:active),
    &:active {
      transform: none;
      box-shadow: none;
      border-color: $davys-gray;
      color: $davys-gray;
    }
  }
}

@mixin button-translations($border-color: $control-color-black, $focus-color) {
  &:hover:not(:active),
  &:focus-visible {
    transform: $control-hover-translation;
  }

  &:focus-visible {
    box-shadow: $control-box-shadow-hover-size $focus-color;
    border-color: $focus-color;
  }

  &:hover:not(:active) {
    box-shadow: $control-box-shadow-hover-size $border-color;
    border-color: $border-color;
  }

  &:active,
  &:focus-visible:active {
    transform: $control-active-translation;
    box-shadow: none;
  }

  &:active {
    border-color: $border-color;
  }

  &:focus-visible:active {
    border-color: $focus-color;
  }
}

.task-list {
  margin: 2 * shared.$step;
  box-shadow: 0px 10px 15px -3px rgba(0, 0, 0, 0.1);
  border-radius: 4px;
  padding: 2 * shared.$step;
  min-width: 740px;

  &__row {
    display: flex;
    column-gap: 2 * shared.$step;
    flex-wrap: nowrap;
    align-items: center;
    white-space: nowrap;
  }

  & > &__row:not(:last-child) {
    margin-bottom: 2 * shared.$step;
  }

  &__button {
    @include control-default-shared-style();
    @include button-translations($focus-color: $cerulean);

    display: block;
    transform: $control-default-translation;
    cursor: pointer;
    box-shadow: $control-box-shadow-default-size $control-color-black;
    background: $control-color-gray;
    font-weight: bold;

    &_remove {
      @include button-translations(
        $border-color: $imperial-red,
        $focus-color: $imperial-red
      );

      background: $control-color-gray no-repeat center
        url('./assets/remove.svg');
      background-size: 80%;
      width: 30px;
      height: 30px;
    }

    &_complete {
      @include button-translations(
        $border-color: $magestic-green,
        $focus-color: $magestic-green
      );

      &:hover,
      &:focus-visible,
      &:active {
        color: $magestic-green;
      }
    }
  }

  &__input {
    @include control-default-shared-style();

    min-width: 300px;

    &:focus {
      transform: $control-default-translation;
      box-shadow: $control-box-shadow-default-size $cerulean;
      border-color: $cerulean;
    }
  }

  &__container {
    display: flex;
    flex-direction: column;
    align-items: center;
  }

  &__item {
    column-gap: 2 * shared.$step;
    max-width: 900px;

    &:not(:last-child) {
      margin-bottom: 2 * shared.$step;
    }

    &_animated {
      @keyframes appear {
        from {
          transform: translateY(0.33em);
          opacity: 0;
        }
      }

      animation: appear $task-item-fade-in-fade-out-animation-length ease;
    }

    &_animated-remove {
      @keyframes fadeOut {
        to {
          transform: translateY(-0.33em);
          opacity: 0;
        }
      }

      animation: fadeOut $task-item-fade-in-fade-out-animation-length ease;
    }
  }

  &__checkbox {
    align-self: center;
  }

  &-item {
    display: flex;
    align-items: center;
    transition: transform ease 0.1s;
    background-color: $platinum;
    padding: 2 * shared.$step;
    width: 100%;

    &:hover {
      transform: scale(1.01);
    }

    &__chevron {
      flex-shrink: 0;
      background: url('./assets/chevron-right.svg') no-repeat center;
      background-size: 100%;
      width: 40px;
      height: 40px;
    }

    &__text {
      flex-grow: 1;
      word-break: break-word;

      &_completed {
        text-decoration: line-through;
      }
    }

    &__controls {
      display: flex;
      row-gap: shared.$step;
      flex-flow: column;
      justify-content: space-between;
      align-items: flex-end;
    }

    &__created-date {
      white-space: nowrap;
    }

    &_completed {
      transition: background-color ease 0.1s;
      background-color: $light-dimmed-green;
    }
  }
}
