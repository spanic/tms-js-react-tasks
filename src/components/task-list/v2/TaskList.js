import { useEffect, useState } from 'react';

import '../TaskList.scss';

import Task from './components/Task';

function TaskList() {
  const [tasks, setTasks] = useState();

  useEffect(() => {
    async function fetchTasks() {
      const tasksObj = await fetch('/tasks/v2').then(
        (response) => response.ok && response.json()
      );
      const tasks = initTasksMap(tasksObj);
      setTasks(tasks);
    }
    fetchTasks();
  }, []);

  function initTasksMap(tasksObj) {
    return new Map(Object.entries(tasksObj));
  }

  async function onComplete(task, key) {
    await fetch(`/tasks/v2/${key}`, {
      method: 'PATCH',
      body: JSON.stringify({ completed: true }),
      headers: {
        'Content-type': 'application/json; charset=UTF-8',
      },
    });
    const completedTaskObj = { ...tasks.get(key), completed: true };
    const tasksCopy = new Map(tasks);
    tasksCopy.set(key, completedTaskObj);
    setTasks(tasksCopy);
  }

  return (
    <div id="task-list" className="task-list">
      <div className="task-list__row">
        <button className="task-list__button">Delete last</button>
        <button className="task-list__button">Delete all</button>
        <input
          id="taskTextInput"
          className="task-list__input"
          type="text"
          placeholder="Enter your task here..."
        ></input>
        <button id="addBtn" className="task-list__button">
          Add
        </button>
      </div>
      <div className="task-list__row">
        <div>
          Total: <span>2</span>
        </div>
        <div>
          Completed: <span>1</span>
        </div>
        <button className="task-list__button">Show all</button>
        <button className="task-list__button" disabled>
          Show completed
        </button>
        <input
          className="task-list__input"
          type="search"
          placeholder="Search..."
          disabled
        ></input>
      </div>

      <div className="task-list__container">
        {tasks &&
          Array.from(tasks.entries()).map(([key, task]) => (
            <Task
              key={key}
              task={task}
              complete={() => onComplete(task, key)}
            />
          ))}
      </div>
    </div>
  );
}

export default TaskList;
