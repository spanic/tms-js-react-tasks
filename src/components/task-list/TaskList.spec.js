/* eslint-disable testing-library/no-node-access */

/* eslint-disable testing-library/no-container */
import { fireEvent, render, screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import TaskList from './TaskList';

// Temporary opted out these tests, should be fixed later
xdescribe('Validates TaskList component', () => {
  test('renders "Delete all button"', () => {
    render(<TaskList />);
    const deleteAllButtonEl = screen.getByText(/delete all/i, {
      selector: 'button.task-list__button',
    });
    expect(deleteAllButtonEl).toBeInTheDocument();
  });

  test('renders "Delete last" button', () => {
    render(<TaskList />);
    const deleteLastButtonEl = screen.getByText(/delete last/i, {
      selector: 'button.task-list__button',
    });
    expect(deleteLastButtonEl).toBeInTheDocument();
  });

  test('renders "Enter your task here..." input', () => {
    render(<TaskList />);
    const enterTaskInputEl =
      screen.getByPlaceholderText(/enter your task here/i);
    expect(enterTaskInputEl).toBeInTheDocument();
  });

  test('renders "Total" block', () => {
    render(<TaskList />);
    const totalBlockEl = screen.getByText(/total:/i);
    expect(totalBlockEl).toBeInTheDocument();
    expect(totalBlockEl).toMatchSnapshot();
  });

  test('renders "Show completed" button', () => {
    render(<TaskList />);
    const showCompletedButton = screen.getByText(/show completed/i);
    expect(showCompletedButton).toBeInTheDocument();
    expect(showCompletedButton).toMatchSnapshot();
  });

  test('renders first task block', () => {
    const { container } = render(<TaskList />);
    const firstTaskBlock = container.querySelector('div.task-list__item');
    expect(firstTaskBlock).toBeInTheDocument();
    expect(firstTaskBlock).toMatchSnapshot();
  });

  it('adds new task', async () => {
    const { container } = render(<TaskList />);
    const user = userEvent.setup();

    const taskItemNameInputEl = container.querySelector('#taskTextInput');
    taskItemNameInputEl.value = 'test';

    const addButton = screen.getByRole('button', {
      name: /add/i,
    });
    user.click(addButton);

    await waitFor(() => {
      return expect(screen.getByText('test')).toBeInTheDocument();
    });
    await waitFor(() => {
      return expect(
        container.querySelectorAll('div.task-list__item').length
      ).toBe(3);
    });
  });

  it('removes all the tasks from the list', async () => {
    const { container } = render(<TaskList />);
    const user = userEvent.setup();

    const deleteAllButton = screen.getByRole('button', {
      name: /delete all/i,
    });
    user.click(deleteAllButton);

    await waitFor(() => {
      return expect(
        container.querySelectorAll('div.task-list__item').length
      ).toBe(0);
    });
  });

  it('removes the selected task from the list', async () => {
    const { container } = render(<TaskList />);
    const user = userEvent.setup();

    const deleteButton = container.querySelector(
      'button.task-list__button_remove'
    );
    await user.click(deleteButton);

    const firstTaskItem = document.querySelector(
      'div.task-list__item_animated-remove'
    );
    // manually triggering 'animationend' event, because Jest doesn't trigger it after assigning the '_animated-remove' class
    fireEvent.animationEnd(firstTaskItem);

    expect(firstTaskItem).not.toBeInTheDocument();
  });

  it('marks the selected task as completed', async () => {
    const { container } = render(<TaskList />);
    const user = userEvent.setup();

    const completeButton = container.querySelector(
      '.task-list__button_complete'
    );
    await user.click(completeButton);

    const completedTaskItem = document.querySelector(
      '.task-list-item_completed .task-list-item__text_completed'
    );
    return expect(completedTaskItem).toBeInTheDocument();
  });
});
