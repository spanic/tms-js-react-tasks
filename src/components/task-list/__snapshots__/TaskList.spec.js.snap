// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Validates TaskList component renders "Show completed" button 1`] = `
<button
  class="task-list__button"
  disabled=""
>
  Show completed
</button>
`;

exports[`Validates TaskList component renders "Total" block 1`] = `
<div>
  Total: 
  <span>
    2
  </span>
</div>
`;

exports[`Validates TaskList component renders first task block 1`] = `
<div
  class="task-list__item task-list-item "
>
  
      
  <div
    class="task-list-item__chevron"
  />
  
      
  <span
    class="task-list-item__text"
  >
    
        Test task
      
  </span>
  
      
  <div
    class="task-list-item__controls"
  >
    
        
    <div
      class="task-list__row"
    >
      
          
      <button
        class="task-list__button task-list__button_complete"
      >
        
            Complete
          
      </button>
      
          
      <button
        class="task-list__button task-list__button_remove"
      />
      
        
    </div>
    
        
    <span
      class="task-list-item__created-date"
    >
      
          02.03.2023 00:13 GMT+6
        
    </span>
    
      
  </div>
  
    
</div>
`;
